[SPARK-169] Re-implement !list & !quote's formatter by using a centralized template

theunkn0wn1 opened the pull request:
This PR addresses [SPARK-169] by introducing `jinja2` for templating, and migrating both `!list` and `!quote`'s handlers to use it.

In basic, the core issue was the two different commands having a different bespoke renderers that do basically the same thing.
I previously took a stab at this in previous PRs, the most recent involving `__format__`. Ultimately im not pleased with that interface, so this is *yet another attempt* at it.

You may be thinking "why jinja" and "isn't jinja for HTML rendering?". To the former, its mainly due to it being a template engine, which makes it better suited to rendering the same stuff (templated stuff) with slight variations than pure python. To the latter, its not just:tm: limited to rendering HTML, it can be used for generating any textual documents. For another project i am using it for code generation (Something this project should hopefully never need....)


# Implementation details
The new render is split into three templates, `board`, `rescue`, `quotation`.
Three templates are used since each one has some special logic, and splitting it up keeps the (relative) complexity lower.

The templates are rendered asynchronously, since jinja supports that.

I ended up inserting a bunch of stuff into the template namespace to facilitate rendering, including functions to render the board & rescues. I also included some general purpose stuff, such as `pendulum.now` and the IRC text formatting functions `bold`, `italic`, `color` and the color enum `Color`.

Rendering of a rescue is primarially controlled by the passed `ListFlags` instance, which has been expanded somewhat to cover most the things the template needs to alter its output accordingly, as to minimize the string manipulation that needs to be done by hand.

Proposed output sample:
![image](https://user-images.githubusercontent.com/3110986/103719435-e071e900-4f7d-11eb-873f-0b2df9a46bea.png)

theunkn0wn1 committed:
Started work on redoing !list

theunkn0wn1 committed:
Improve rat rendering

theunkn0wn1 committed:
Move stuff from scratch file into project namespace

theunkn0wn1 committed:
Oops, testing code snuck in

theunkn0wn1 committed:
does turn out there was an easier way of doing the for loops. does help to RTFM!

theunkn0wn1 committed:
Further work towards rendering rescues via jinja

theunkn0wn1 committed:
Minor whitespace things. grey out the api ID as its less "important"

theunkn0wn1 committed:
Start work on rendering quotations via templates

Master-Guy commented:
[Error] Please rebase to develop before submitting the PR
[Testing] No unit tests?
[Consensus]/[Discussion] The output looks to be messy - Sometimes, but not always, empty lines after a rescue

```
[02-01-2021 10:14:30] <@Master-Guy> !inject MG CR 00:00 PC SOL
[02-01-2021 10:14:30] <@Master-Guy> !inject MG2 00:00 PC
[02-01-2021 10:14:30] <@Master-Guy> !inject MG3 PC
[02-01-2021 10:14:30] <@Master-Guy> !inject MG4 PS
[02-01-2021 10:14:30] <@Master-Guy> !assign 1 test_rat
[02-01-2021 10:14:30] <@Master-Guy> !active 2 client left

[02-01-2021 10:14:49] <@Master-Guy> !list -irc
[02-01-2021 10:14:49] <Mecha-Guy[BOT]> [0] MG's case (PC) in  <unspecified> 
[02-01-2021 10:14:49] <Mecha-Guy[BOT]>         No assigned Rats.  
[02-01-2021 10:14:49] <Mecha-Guy[BOT]> [3] MG4's case (PS) in  <unspecified> 
[02-01-2021 10:14:49] <Mecha-Guy[BOT]>         No assigned Rats.  
[02-01-2021 10:14:49] <Mecha-Guy[BOT]> [1] MG2's case (PC) in  <unspecified> 
[02-01-2021 10:14:49] <Mecha-Guy[BOT]>         Assigned Rats: 
[02-01-2021 10:14:49] <Mecha-Guy[BOT]>     Unidentified Rats: test_rat
[02-01-2021 10:14:49] <Mecha-Guy[BOT]>  
[02-01-2021 10:14:49] <Mecha-Guy[BOT]> [2] MG3's case (PC) in  <unspecified>  (Inactive) 
[02-01-2021 10:14:49] <Mecha-Guy[BOT]>         No assigned Rats.  
[02-01-2021 10:14:49] <Mecha-Guy[BOT]>  
```

theunkn0wn1 committed:
Update poetry lock

theunkn0wn1 committed:
Add render call for quotes
 - add pendulum.now to template namespace
 - tie rescue quotes into rescue template

theunkn0wn1 committed:
Tie !quote into this madness

theunkn0wn1 committed:
Make PS stand out more, also bold the platform.

theunkn0wn1 committed:
force UTC time zone when rendering quotes

theunkn0wn1 committed:
Re-do the rescue renderer to make rat rendering easier
 - insert bold and italic into template namespace

theunkn0wn1 committed:
Black pass over template dir

theunkn0wn1 committed:
minor ordering nit.

theunkn0wn1 committed:
Address some linting nits.
Expose ListFlags on the public interface

theunkn0wn1 committed:
Move `ListFlags` -> `RescueRenderFlags`
 - also move where it lives, since its now not necessarially part of command but part of the template

theunkn0wn1 commented:
Updated against current develop.
Relevant tests are already implemented, no need to test precise output when the exact output may change down the line.
The existing tests ensure the relevant information is there.

Trailing newlines should be handled  now, mainly by calling `rstrip('\n')` after the renderer. less effort than figuring out where trailing white space is emitted from the templates.

theunkn0wn1 commented on src/commands/__init__.py:
```diff
@@ -22,4 +23,5 @@
     "starsystems",
     "administration",
     "api_utilities",
+    "RescueRenderFlags"
```
[error] this isn't part of command's namespace now.

theunkn0wn1 commented on src/commands/case_management.py:
```diff
@@ -40,6 +40,7 @@
 from ..packages.rat import Rat
 from ..packages.rescue import Rescue
 from ..packages.utils import Platforms, Status, color, bold, Colors
+from ..templates import env
```
[style] merge this with the import on line 23.

theunkn0wn1 commented on src/templates/render_flags.py:
```diff
@@ -41,4 +43,6 @@ def from_word(cls, argument: str):
         show API UUIDs
         """
-        return cls(show_inactive, filter_unassigned_rescues, show_assigned_rats, show_uuids)
+        show_unident = True  # fixme remove hardcode
```
FIXME

theunkn0wn1 committed:
Resolve FIXME

theunkn0wn1 committed:
Purge `RescueRenderFlags` from `src.commands` Namespace

theunkn0wn1 committed:
rename env -> `template_environment`
 - readability

theunkn0wn1 committed:
Fix imports

theunkn0wn1 commented:
Reviewing SPARK-169, both the inject handler and ~~grab~~ (not as of SPARK-199) should also use these new renderers.

theunkn0wn1 commented:
`!grab` doesn't require attention (thanks to SPARK-199), and `inject` is differed to SPARK-248.

Master-Guy commented:
LGTM 👍 
Last style comment, but very non-blocking: The last rescue also gets an empty line at the end, which seems to be a little bit overkill. Removing this shouldn't affect a lot of things.
